util/semaphore-group/semaphore_group.go

```diff
@@ -0,0 +1,39 @@
+package semaphoregroup
+
+import (
+	"sync"
+)
+
+// SemaphoreGroup is a custom type that combines sync.WaitGroup and a semaphore.
+type SemaphoreGroup struct {
+	waitGroup sync.WaitGroup
+	semaphore chan struct{}
+}
+
+// NewSemaphoreGroup creates a new SemaphoreGroup with the specified semaphore limit.
+func NewSemaphoreGroup(limit int) *SemaphoreGroup {
+	return &SemaphoreGroup{
+		semaphore: make(chan struct{}, limit),
+	}
+}
+
+// Add increments the internal WaitGroup counter and acquires a semaphore slot.
+func (sg *SemaphoreGroup) Add() {
+	sg.waitGroup.Add(1)
+
+	// Acquire semaphore slot
+	sg.semaphore <- struct{}{}
```

Comment:
Maybe this line can cause an infinitie block (depending on how do we use this struct). Maybe it would be safer if we pass a context to the Add(ctx context.Context) and handle the cancel event.